src/isl.rs

```diff
@@ -0,0 +1,211 @@
+                            value.ion_type()
+                        )));
+                    }
+                    let types: Vec<IslTypeRef> = value
+                        .as_sequence()
+                        .unwrap()
+                        .iter()
+                        .map(|e| IslTypeRef::parse_from_ion_element(e))
+                        .collect::<IonSchemaResult<Vec<IslTypeRef>>>()?;
+                    IslConstraint::AllOf(types)
+                }
+                "type" => {
+                    if value.ion_type() != IonType::Symbol && value.ion_type() != IonType::Struct {
+                        return Err(invalid_schema_error_raw(format!(
+                            "type constraint was a {:?} instead of a symbol/struct",
+                            value.ion_type()
+                        )));
+                    }
+                    let type_reference: IslTypeRef = IslTypeRef::parse_from_ion_element(value)?;
+                    IslConstraint::Type(type_reference)
+                }
+                _ => {
+                    return Err(invalid_schema_error_raw(
+                        "Type: ".to_owned()
+                            + &type_name.unwrap()
+                            + " can not be built as constraint: "
+                            + constraint_name
+                            + " does not exist",
+                    ))
+                }
+            };
+            constraints.push(constraint);
+        }
+        Ok(IslType::new(type_name.to_owned(), constraints))
+    }
+}
+
+/// Provides an internal representation of schema type reference.
+/// Type reference grammar is defined in [Ion Schema Spec]
+/// [Ion Schema spec]: https://amzn.github.io/ion-schema/docs/spec.html#grammar
+#[derive(Debug, Clone)]
+pub enum IslTypeRef {
+    /// represents core ion type reference
+    IslCoreType(IonType),
+    /// represents a type reference which represents a type imported from another schema
+    AliasType(String),
+    /// represents a type reference defined as an inlined import type from another schema
+    // TODO: add ImportType(Import) where ImportType could either point to a schema represented by an id with all the types or a single type from inside it
+    /// represents an unnamed type definition reference
+    AnonymousType(IslType),
+}
+
+// TODO: add a check for nullable type reference
+impl IslTypeRef {
+    /// Tries to create an [IslTypeRef] from the given OwnedElement
+    pub fn parse_from_ion_element(value: &OwnedElement) -> IonSchemaResult<Self> {
+        match value.ion_type() {
+            IonType::Symbol => {
+                value.as_sym().unwrap()
+                    .text()
+                    .ok_or_else(|| {
+                        invalid_schema_error_raw(
+                            "a base or alias type reference symbol doesn't have text",
+                        )
+                    })
+                    .and_then(|type_reference| {
```

Comment:
We're in the process of creating a type reference, but I think this particular variable is a type name.
```suggestion
                    .and_then(|type_name| {
```